### PR TITLE
new connector for Atlassian Crowd

### PR DESCRIPTION
**Pull request by diafour**
New connector for Atlassian Crowd identity management server. This initial implementation uses PasswordConnector ans should support groups and refresh. But there is no tests yet and sso feature is not tested with Jira or Confluence. Is this looks interesting?

### CODE REVIEW

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+	ExpiryDate  uint64 `json:"expiry-date"`
+}
+
+type crowdAuthenticationError struct {
+	Reason  string
+	Message string
+}
+
+// Open returns a strategy for logging in through Atlassian Crowd.
+func (c *Config) Open(id string, logger log.Logger) (connector.Connector, error) {
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: [nit] if `id` isn't used, let's do this:
```suggestion
func (c *Config) Open(_ string, logger log.Logger) (connector.Connector, error) {
```

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+	return &crowdConnector{
+		Config: *c,
+		logger: logger,
+	}, nil
+}
+
+type connectorData struct {
+	// GitLab's OAuth2 tokens never expire. We don't need a refresh token.
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: ...GitLab? 😉

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+	User     crowdUser `json:"user"`
+}
+
+func (c *crowdConnector) Login(ctx context.Context, s connector.Scopes, username, password string) (ident connector.Identity, validPass bool, err error) {
+	// make this check to avoid empty passwords.
+	if password == "" {
+		return connector.Identity{}, false, nil
+	}
+
+	c.logger.Infof("crowd Login scopes: %+v", s)
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: Can we remove this or change it to debug? I'm not sure I'd want to see this for every login using this connector 😄

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: ❓ Why do we need this? Does it differ from `http.DefaultTransport`?

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+	}
+}
+
+// authenticateWithPassword creates a new session for user and validates a password with Crowd API
+func (c *crowdConnector) authenticateWithPassword(ctx context.Context, client *http.Client, username string, password string) (invalidPass bool, err error) {
+	req, err := c.crowdUserManagementRequest(ctx,
+		"POST",
+		fmt.Sprintf("/session"),
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]:
```suggestion
		"/session",
```

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+
+	if resp.StatusCode != http.StatusCreated {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return false, fmt.Errorf("crowd: read user body: %v", err)
+		}
+
+		var authError crowdAuthenticationError
+
+		c.logger.Infof("body: %s", string(body))
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: This can also go away I think.

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+	"github.com/dexidp/dex/connector"
+	"github.com/dexidp/dex/pkg/groups"
+	"github.com/dexidp/dex/pkg/log"
+)
+
+const (
+	// read operations of the /api/v4/user endpoint
+	scopeUser = "read_user"
+	// used to retrieve groups from /oauth/userinfo
+	// https://docs.gitlab.com/ee/integration/openid_connect_provider.html
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: gitlab? 😉

File: connector/atlassian_crowd/atlassian_crowd.go [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+	scopeOpenID = "openid"
+)
+
+// Config holds configuration options for Atlassian Crowd conector.
```
Comment (srenatus) [inline connector/atlassian_crowd/atlassian_crowd.go]: [typo] ```suggestion
// Config holds configuration options for Atlassian Crowd connector.
```

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 7 above)
Comment (diafour) [reply connector/atlassian_crowd/atlassian_crowd.go]: This plugin is based on gitlab plugin. Definitely code cleaning is required.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 4 above)
Comment (diafour) [reply connector/atlassian_crowd/atlassian_crowd.go]: It is from a github plugin: https://github.com/flant/dex/blob/51f50fcad87b7ce3b08507f5667b59a327e885b3/connector/github/github.go#L220

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 1 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Thanks, fixed!

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 2 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Sorry, connectorData is not used in the Crowd connector. I removed it.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 3 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Fixed.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 4 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: There is a difference with DefaultTransport in one field - `ForceAttemptHTTP2`.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 5 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Fixed.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 6 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: I got rid of it.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 8 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Fixed.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 7 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Fixed.

File: connector/atlassian_crowd/atlassian_crowd.go (hunk 8 above)
Comment (nabokihms) [reply connector/atlassian_crowd/atlassian_crowd.go]: Fixed

**Comment by nabokihms** [comment]
@srenatus can you take a look at the PR again?

We run the Atlassian Crowd connector in production for a couple of months and everything seems fine!
For now, we:
* Fixed all things you mentioned in this PR
* Wrote the doc about Atlassian Crowd connector
* Added unit tests
* Added SVG icon and fixed CSS

If `@srenatus` is not available, maybe @bonifaido or @sagikazarmark can review this PR?

Thanks in advance!

**Comment by mvdkleijn** [comment]
Is there anything that still needs to be done for this apart from someone giving it a positive review?

**Comment by mvdkleijn** [comment]
@srenatus @bonifaido @sagikazarmark - I just setup dex with this connector and it works beautifully! I don't suppose any of you can give it a review so it can be included in the official dex builds? :-)

**Comment by mvdkleijn** [comment]
Yay! Thanks guys! :smile:

**Pull request merged**